contrib/pkg/awsprivatelink/enable.go

```diff
@@ -0,0 +1,224 @@
+
+	// AWS EC2 client in Hive cluster's region
+	EC2Client *ec2.EC2
+}
+
+func NewEnableAWSPrivateLinkCommand() *cobra.Command {
+	opt := &enableOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "enable",
+		Short: "Enable AWS PrivateLink",
+		Long: `Enable AWS PrivateLink:
+1) Extract AWS Hub account credentials from the environment where this command is called
+2) Create a Secret with the above credential. The Secret is referred to by HiveConfig.spec.awsPrivateLink.credentialsSecretRef
+3) Add the active cluster's VPC to the list of HiveConfig.spec.awsPrivateLink.associatedVPCs`,
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := opt.Complete(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Validate(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Run(cmd, args); err != nil {
+				return
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&opt.dnsRecordType, "dns-record-type", "Alias", "HiveConfig.spec.awsPrivateLink.dnsRecordType")
+	return cmd
+}
+
+func (o *enableOptions) Complete(cmd *cobra.Command, args []string) error {
+	// Get client config
+	cfg, err := hiveutils.GetClientConfig()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to generate client config")
+	}
+
+	// Get Kubernetes client
+	kubernetesClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create Kubernetes clientset")
+	}
+	o.kubernetesClient = kubernetesClient
+
+	// Get Hive client
+	hiveClient, err := hiveclient.NewForConfig(cfg)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create Hive clientset")
+	}
+	o.hiveClient = hiveClient
+
+	// Get Openshift CoreV1 client
+	openshiftConfigClient, err := openshiftcorev1.NewForConfig(cfg)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create Openshift Config/v1 client")
+	}
+	o.openshiftConfigClient = openshiftConfigClient
+
+	// Get current user's home directory
+	o.homeDir = "."
+	u, err := user.Current()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to get the current user's home directory")
+	}
+	o.homeDir = u.HomeDir
+
+	// Get EC2 client
+	o.region, o.infraId = o.getRegionAndInfraId()
+	log.Debugf("Found region = %v, infraId = %v", o.region, o.infraId)
+	o.EC2Client = awsutils.GetAWSEC2Client(o.region)
+
+	return nil
+}
+
+func (o *enableOptions) Validate(cmd *cobra.Command, args []string) error {
+	switch o.dnsRecordType {
+	case "Alias", "ARecord":
+	default:
+		log.Fatal(`--dns-record-type must be one of "Alias", "ARecord"`)
+	}
+
+	return nil
+}
+
+func (o *enableOptions) Run(cmd *cobra.Command, args []string) error {
+	// Get HiveConfig, make sure AWS PrivateLink is not enabled
+	hiveConfig, err := o.hiveClient.HiveV1().HiveConfigs().Get(context.Background(), "hive", metav1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatal("Failed to get HiveConfig/hive")
+	}
+	if hiveConfig.Spec.AWSPrivateLink != nil {
+		log.Fatal("AWS Private Link is already enabled")
+	}
```

Comment:
I'm wondering if it's worth proceeding for the sake of idempotence. I.e. if something went wrong partway through the last time we invoked this, it would be nice to be able to (fix the problem and) run it again to complete the configuration.

Reply:
I'm hesitating between the following options:
1) Abandon immediately in case HiveConfig.spec.awsPrivateLink != nil (current implementation)
2) Proceed regardless of the value of HiveConfig.spec.awsPrivateLink. In this way the existing HiveConfig.spec.awsPrivateLink is re-written.
3) Proceed regardless of the value of HiveConfig.spec.awsPrivateLink, but do not touch HiveConfig.spec.awsPrivateLink if it is non-nil (similar to the current implementation)

Reply:
I was thinking somewhere between 2 and 3: populate each field iff it is unset. That may be more work than it's worth though.

How about we leave it as is, but beef up the failure message:
```suggestion
	if hiveConfig.Spec.AWSPrivateLink != nil {
		log.Fatal("AWS Private Link is already enabled. If a previous configuration attempt did not complete, you can clean up via `hiveutil awsprivatelink disable` and try again.")
	}
```
As long as `disable` is thorough and idempotent, this advice should be sound, yes?

Reply:
Yep. Sounds reasonable to me.